File: src/nova_audio_agent/realtime/protocol.py
"""Provider-neutral realtime session commands and normalized events."""

from __future__ import annotations

from collections.abc import AsyncIterator
from dataclasses import dataclass
from typing import Any, Literal, Protocol, TypeAlias

from nova_audio_agent.realtime.history import MAX_PACKED_RECOVERY_CONTENT

MAX_REALTIME_TEXT = 4000
HostItemKind = Literal["progress", "final", "recovery", "dialogue_context", "tool_output"]
HostResponseKind = Literal["host_fact", "tool_result", "delegation_acknowledgement"]
ResponseStatus = Literal["completed", "cancelled", "failed"]
ResponseCancelRejectReason = Literal["no_active_response"]


class ItemDeliveryUncertainError(RuntimeError):
    """The provider may have accepted an item whose confirmation never arrived."""

    def __init__(
        self,
        *,
        session_epoch: int,
        host_item_id: str,
        provider_item_id: str,
        item_kind: HostItemKind,
    ) -> None:
        _require_epoch(session_epoch)
        _require_id(host_item_id, "host_item_id")
        _require_id(provider_item_id, "provider_item_id")
        if item_kind not in {
            "progress",
            "final",
            "recovery",
            "dialogue_context",
            "tool_output",
        }:
            raise ValueError("unknown host item kind")
        super().__init__("host item confirmation timed out; delivery is uncertain")
        self.session_epoch = session_epoch
        self.host_item_id = host_item_id
        self.provider_item_id = provider_item_id
        self.item_kind = item_kind


def _require_id(value: str, field: str) -> None:
    if type(value) is not str or not value.strip():
        raise ValueError(f"{field} must be a non-empty string")


def _require_epoch(value: int) -> None:
    if type(value) is not int or value < 1:
        raise ValueError("session_epoch must be a positive integer")


def _require_text(value: str, field: str, *, allow_empty: bool = False) -> None:
    if type(value) is not str or (not allow_empty and not value.strip()):
        raise ValueError(f"{field} must be a non-empty string")
    if len(value) > MAX_REALTIME_TEXT:
        raise ValueError(f"{field} exceeds {MAX_REALTIME_TEXT} characters")


@dataclass(frozen=True, slots=True)
class HostContextItem:
    kind: HostItemKind
    host_item_id: str
    event_id: str
    content: str
    call_id: str | None = None

    def __post_init__(self) -> None:
        if self.kind not in {
            "progress",
            "final",
            "recovery",
            "dialogue_context",
            "tool_output",
        }:
            raise ValueError("unknown host item kind")
        _require_id(self.host_item_id, "host_item_id")
        _require_id(self.event_id, "event_id")
        _require_text(self.content, "content")
        if self.kind == "dialogue_context" and len(self.content) > MAX_PACKED_RECOVERY_CONTENT:
            raise ValueError(f"dialogue context exceeds {MAX_PACKED_RECOVERY_CONTENT} characters")
        if self.kind == "tool_output":
            if self.call_id is None:
                raise ValueError("tool output call_id is required")
            _require_id(self.call_id, "call_id")
        elif self.call_id is not None:
            raise ValueError("call_id is only valid for tool output")

    @classmethod
    def progress(cls, *, host_item_id: str, event_id: str, content: str) -> HostContextItem:
        return cls(
            kind="progress",
            host_item_id=host_item_id,
            event_id=event_id,
            content=content,
        )

    @classmethod
    def final(cls, *, host_item_id: str, event_id: str, content: str) -> HostContextItem:
        return cls(kind="final", host_item_id=host_item_id, event_id=event_id, content=content)

    @classmethod
    def recovery(cls, *, host_item_id: str, event_id: str, content: str) -> HostContextItem:
        return cls(kind="recovery", host_item_id=host_item_id, event_id=event_id, content=content)

    @classmethod
    def dialogue_context(cls, *, host_item_id: str, event_id: str, content: str) -> HostContextItem:
        return cls(
            kind="dialogue_context",
            host_item_id=host_item_id,
            event_id=event_id,
            content=content,
        )

    @classmethod
    def tool_output(
        cls,
        *,
        host_item_id: str,
        event_id: str,
        call_id: str,
        content: str,
    ) -> HostContextItem:
        return cls(
            kind="tool_output",
            host_item_id=host_item_id,
            event_id=event_id,
            content=content,
            call_id=call_id,
        )


@dataclass(frozen=True, slots=True)
class HostResponseIntent:
    kind: HostResponseKind
    item: HostContextItem
    task_summary: str | None = None
    # #55: host-asserted fact that the origin response already voiced this
    # acceptance; the provider renders a hard silence directive instead of the
    # self-judged soft clause.
    origin_spoken: bool = False

    def __post_init__(self) -> None:
        if self.kind not in {"host_fact", "tool_result", "delegation_acknowledgement"}:
            raise ValueError("unknown host response kind")
        if type(self.origin_spoken) is not bool:
            raise ValueError("origin_spoken must be a bool")
        if self.kind == "delegation_acknowledgement":
            if self.item.kind != "tool_output":
                raise ValueError("delegation acknowledgement requires tool output")
            if self.task_summary is None:
                raise ValueError("task_summary is required")
            _require_text(self.task_summary, "task_summary")
        else:
            if self.task_summary is not None:
                raise ValueError("task_summary is only valid for delegation acknowledgement")
            if self.origin_spoken:
                raise ValueError("origin_spoken is only valid for delegation acknowledgement")

    @classmethod
    def host_fact(cls, item: HostContextItem) -> HostResponseIntent:
        if item.kind == "tool_output":
            raise ValueError("host fact cannot be tool output")
        if item.kind == "dialogue_context":
            raise ValueError("dialogue context cannot create a response")
        return cls(kind="host_fact", item=item)

    @classmethod
    def tool_result(cls, item: HostContextItem) -> HostResponseIntent:
        if item.kind != "tool_output":
            raise ValueError("tool result requires tool output")
        return cls(kind="tool_result", item=item)

    @classmethod
    def delegation_acknowledgement(
        cls,
        *,
        item: HostContextItem,
        task_summary: str,
        origin_spoken: bool = False,
    ) -> HostResponseIntent:
        return cls(
            kind="delegation_acknowledgement",
            item=item,
            task_summary=task_summary,
            origin_spoken=origin_spoken,
        )


@dataclass(frozen=True, slots=True)
class SessionIdentity:
    epoch: int
    provider_session_id: str

    def __post_init__(self) -> None:
        _require_epoch(self.epoch)
        _require_id(self.provider_session_id, "provider_session_id")


@dataclass(frozen=True, slots=True)
class ItemIdentity:
    session_epoch: int
    host_item_id: str
    provider_item_id: str

    def __post_init__(self) -> None:
        _require_epoch(self.session_epoch)
        _require_id(self.host_item_id, "host_item_id")
        _require_id(self.provider_item_id, "provider_item_id")


@dataclass(frozen=True, slots=True)
class _SessionEvent:
    session_epoch: int

    def __post_init__(self) -> None:
        _require_epoch(self.session_epoch)


@dataclass(frozen=True, slots=True)
class UserSpeechStarted(_SessionEvent):
    speech_id: str
    provider_item_id: str | None = None

    def __post_init__(self) -> None:
        super(UserSpeechStarted, self).__post_init__()
        _require_id(self.speech_id, "speech_id")
        if self.provider_item_id is not None:
            _require_id(self.provider_item_id, "provider_item_id")


@dataclass(frozen=True, slots=True)
class UserSpeechEnded(_SessionEvent):
    speech_id: str
    provider_item_id: str | None = None

    def __post_init__(self) -> None:
        super(UserSpeechEnded, self).__post_init__()
        _require_id(self.speech_id, "speech_id")
        if self.provider_item_id is not None:
            _require_id(self.provider_item_id, "provider_item_id")


@dataclass(frozen=True, slots=True)
class _ItemTextEvent(_SessionEvent):
    item_id: str
    text: str

    def __post_init__(self) -> None:
        super(_ItemTextEvent, self).__post_init__()
        _require_id(self.item_id, "item_id")
        _require_text(self.text, "text", allow_empty=True)


@dataclass(frozen=True, slots=True)
class UserTranscriptDelta(_ItemTextEvent):
    pass


@dataclass(frozen=True, slots=True)
class UserTranscriptFailed(_SessionEvent):
    item_id: str

    def __post_init__(self) -> None:
        super(UserTranscriptFailed, self).__post_init__()
        _require_id(self.item_id, "item_id")


@dataclass(frozen=True, slots=True)
class UserTranscriptFinal(_ItemTextEvent):
    pass


@dataclass(frozen=True, slots=True)
class ResponseStarted(_SessionEvent):
    response_id: str

    def __post_init__(self) -> None:
        super(ResponseStarted, self).__post_init__()
        _require_id(self.response_id, "response_id")


@dataclass(frozen=True, slots=True)
class ResponseAudioDelta:
    session_epoch: int
    response_id: str
    pcm: bytes

    def __post_init__(self) -> None:
        _require_epoch(self.session_epoch)
        _require_id(self.response_id, "response_id")
        if type(self.pcm) is not bytes or not self.pcm or len(self.pcm) % 2:
            raise ValueError("pcm must be non-empty aligned PCM16 bytes")


@dataclass(frozen=True, slots=True)
class _ResponseTextEvent(_SessionEvent):
    response_id: str
    text: str

    def __post_init__(self) -> None:
        super(_ResponseTextEvent, self).__post_init__()
        _require_id(self.response_id, "response_id")
        _require_text(self.text, "text", allow_empty=True)


@dataclass(frozen=True, slots=True)
class ResponseTranscriptDelta(_ResponseTextEvent):
    pass


@dataclass(frozen=True, slots=True)
class ResponseTranscriptFinal(_ResponseTextEvent):
    pass


@dataclass(frozen=True, slots=True)
class ToolCallReady(_SessionEvent):
    call_id: str
    item_id: str
    name: str
    arguments: dict[str, Any]
    response_id: str | None = None

    def __post_init__(self) -> None:
        super(ToolCallReady, self).__post_init__()
        _require_id(self.call_id, "call_id")
        _require_id(self.item_id, "item_id")
        _require_id(self.name, "name")
        if type(self.arguments) is not dict:
            raise ValueError("arguments must be an object")
        if self.response_id is not None:
            _require_id(self.response_id, "response_id")


@dataclass(frozen=True, slots=True)
class ItemConfirmed(_SessionEvent):
    host_item_id: str
    provider_item_id: str

    def __post_init__(self) -> None:
        super(ItemConfirmed, self).__post_init__()
        _require_id(self.host_item_id, "host_item_id")
        _require_id(self.provider_item_id, "provider_item_id")


@dataclass(frozen=True, slots=True)
class ResponseTerminal(_SessionEvent):
    response_id: str
    status: ResponseStatus
    reason: str

    def __post_init__(self) -> None:
        super(ResponseTerminal, self).__post_init__()
        _require_id(self.response_id, "response_id")
        if self.status not in {"completed", "cancelled", "failed"}:
            raise ValueError("unknown response status")
        _require_text(self.reason, "reason")


@dataclass(frozen=True, slots=True)
class ResponseCancelRejected(_SessionEvent):
    response_id: str
    cancel_request_id: str
    reason: ResponseCancelRejectReason

    def __post_init__(self) -> None:
        super(ResponseCancelRejected, self).__post_init__()
        _require_id(self.response_id, "response_id")
        _require_id(self.cancel_request_id, "cancel_request_id")
        if self.reason != "no_active_response":
            raise ValueError("unknown response cancel rejection reason")


@dataclass(frozen=True, slots=True)
class ProviderErrorEvent(_SessionEvent):
    code: str
    recoverable: bool = False

    def __post_init__(self) -> None:
        super(ProviderErrorEvent, self).__post_init__()
        _require_id(self.code, "code")
        if type(self.recoverable) is not bool:
            raise ValueError("recoverable must be bool")


RealtimeFrontBrainEvent: TypeAlias = (
    UserSpeechStarted
    | UserSpeechEnded
    | UserTranscriptDelta
    | UserTranscriptFailed
    | UserTranscriptFinal
    | ResponseStarted
    | ResponseAudioDelta
    | ResponseTranscriptDelta
    | ResponseTranscriptFinal
    | ToolCallReady
    | ItemConfirmed
    | ResponseTerminal
    | ResponseCancelRejected
    | ProviderErrorEvent
)


class RealtimeFrontBrain(Protocol):
    async def connect(self, *, tools: tuple[dict[str, Any], ...]) -> SessionIdentity: ...

    async def send_audio(self, pcm: bytes) -> None: ...

    async def inject_host_item(
        self,
        item: HostContextItem,
        *,
        confirmation_timeout: float | None = None,
        as_user_activation: bool = False,
    ) -> ItemIdentity: ...

    async def create_response(self, intent: HostResponseIntent) -> None: ...

    async def cancel_response(self, response_id: str) -> None: ...

    def events(self) -> AsyncIterator[RealtimeFrontBrainEvent]: ...

    async def close(self) -> None: ...

File: runtime/src/realtime/protocol.ts
import { z } from 'zod'
import { jsonValueSchema, type JsonValue } from '../events.js'
import { MAX_PACKED_RECOVERY_CONTENT } from './history.js'

export const MAX_REALTIME_TEXT = 4_000
export const MAX_REALTIME_PCM_BYTES = 64 * 1_024

const pythonWhitespaceOnly = /^[\u0009-\u000d\u001c-\u0020\u0085\u00a0\u1680\u2000-\u200a\u2028\u2029\u202f\u205f\u3000]*$/u
export const realtimeIdentifierSchema = z.string().refine(
  value => !pythonWhitespaceOnly.test(value),
  'identifier must be non-empty',
)
const epochSchema = z.number().int().positive()
const jsonObjectSchema = z.record(z.string(), jsonValueSchema)

function boundedText(allowEmpty = false): z.ZodString {
  return z.string()
    .refine(value => allowEmpty || !pythonWhitespaceOnly.test(value), 'text must be non-empty')
    .refine(value => [...value].length <= MAX_REALTIME_TEXT,
      `text exceeds ${MAX_REALTIME_TEXT} characters`)
}

export const hostItemKindSchema = z.enum([
  'progress',
  'final',
  'recovery',
  'dialogue_context',
  'tool_output',
  'workspace_context',
])

export const hostContextItemSchema = z.object({
  kind: hostItemKindSchema,
  host_item_id: realtimeIdentifierSchema,
  event_id: realtimeIdentifierSchema,
  content: boundedText(),
  call_id: realtimeIdentifierSchema.nullable().default(null),
  session_epoch: epochSchema.optional(),
  workspace_instance_id: realtimeIdentifierSchema.optional(),
  revision: z.number().int().nonnegative().optional(),
}).strict().superRefine((item, context) => {
  if (
    item.kind === 'dialogue_context'
    && [...item.content].length > MAX_PACKED_RECOVERY_CONTENT
  ) {
    context.addIssue({
      code: 'custom',
      path: ['content'],
      message: `dialogue context exceeds ${MAX_PACKED_RECOVERY_CONTENT} characters`,
    })
  }
  if (item.kind === 'tool_output' && item.call_id === null) {
    context.addIssue({code: 'custom', path: ['call_id'], message: 'tool output call_id is required'})
  } else if (item.kind !== 'tool_output' && item.call_id !== null) {
    context.addIssue({
      code: 'custom',
      path: ['call_id'],
      message: 'call_id is only valid for tool output',
    })
  }
  const workspaceContextFields = [
    item.session_epoch,
    item.workspace_instance_id,
    item.revision,
  ]
  if (item.kind === 'workspace_context') {
    for (const [index, field] of workspaceContextFields.entries()) {
      if (field === undefined) {
        context.addIssue({
          code: 'custom',
          path: [['session_epoch', 'workspace_instance_id', 'revision'][index]!],
          message: 'workspace context requires session_epoch, workspace_instance_id, and revision',
        })
      }
    }
  } else if (workspaceContextFields.some(field => field !== undefined)) {
    context.addIssue({
      code: 'custom',
      message: 'workspace context fields are only valid for workspace_context',
    })
  }
})

export type HostContextItem = z.infer<typeof hostContextItemSchema>

/**
 * A provider adapter reports one of these capabilities before it is allowed to make a
 * workspace-context item provider-visible. This is a contract declaration, not proof that a
 * specific provider implements the operation; provider-specific proof belongs at the adapter.
 */
export const workspaceContextDeliveryCapabilitySchema = z.enum([
  'replace_provider_item',
  'refresh_session',
  'unavailable',
])

/**
 * The provider-visible transition for a workspace-context revision. A previous provider item is
 * never left visible while a newer revision is accepted: it is either superseded explicitly or a
 * session refresh is recorded. `unavailable` is intentionally a non-delivery state.
 */
const workspaceContextDeliveryIdentity = {
  session_epoch: epochSchema,
  workspace_instance_id: realtimeIdentifierSchema,
  revision: z.number().int().nonnegative(),
  // The adapter must attest to the prior provider-visible state, even when it was empty.
  prior_provider_item_id: realtimeIdentifierSchema.nullable(),
}

const replaceWorkspaceContextDeliverySchema = z.object({
  capability: z.literal('replace_provider_item'),
  delivered: z.literal(true),
  ...workspaceContextDeliveryIdentity,
  provider_item_id: realtimeIdentifierSchema,
  superseded_provider_item_id: realtimeIdentifierSchema.nullable(),
}).strict().superRefine((delivery, context) => {
  if (delivery.superseded_provider_item_id !== delivery.prior_provider_item_id) {
    context.addIssue({
      code: 'custom',
      path: ['superseded_provider_item_id'],
      message: 'prior provider state must be explicitly superseded before a newer revision is visible',
    })
  }
  if (
    delivery.prior_provider_item_id !== null
    && delivery.provider_item_id === delivery.prior_provider_item_id
  ) {
    context.addIssue({
      code: 'custom',
      path: ['provider_item_id'],
      message: 'newer revision requires a distinct provider item id',
    })
  }
})

const refreshWorkspaceContextDeliverySchema = z.object({
  capability: z.literal('refresh_session'),
  delivered: z.literal(true),
  ...workspaceContextDeliveryIdentity,
  refresh_id: realtimeIdentifierSchema,
}).strict()

const unavailableWorkspaceContextDeliverySchema = z.object({
  capability: z.literal('unavailable'),
  delivered: z.literal(false),
  ...workspaceContextDeliveryIdentity,
}).strict()

export const workspaceContextDeliverySchema = z.discriminatedUnion('capability', [
  replaceWorkspaceContextDeliverySchema,
  refreshWorkspaceContextDeliverySchema,
  unavailableWorkspaceContextDeliverySchema,
])

export const workspaceContextDeliveryRecordSchema = z.object({
  item: hostContextItemSchema,
  asUserActivation: z.literal(false),
  delivery: workspaceContextDeliverySchema,
}).strict().superRefine((injection, context) => {
  if (injection.item.kind !== 'workspace_context') {
    context.addIssue({
      code: 'custom',
      path: ['item'],
      message: 'workspace context injection requires workspace_context item kind',
    })
  }
  if (
    injection.item.session_epoch !== injection.delivery.session_epoch
    || injection.item.workspace_instance_id !== injection.delivery.workspace_instance_id
    || injection.item.revision !== injection.delivery.revision
  ) context.addIssue({
    code: 'custom',
    path: ['delivery'],
    message: 'workspace context delivery proof must bind the exact session_epoch, workspace_instance_id, and revision',
  })
})

export const workspaceContextInjectionSchema = workspaceContextDeliveryRecordSchema.superRefine(
  (injection, context) => {
    if (injection.delivery.capability === 'unavailable') {
      context.addIssue({
        code: 'custom',
        path: ['delivery', 'capability'],
        message: 'workspace context is unavailable and cannot be injected',
      })
    }
  },
)

export type WorkspaceContextDeliveryCapability = z.infer<typeof workspaceContextDeliveryCapabilitySchema>
export type WorkspaceContextDelivery = z.infer<typeof workspaceContextDeliverySchema>
export type WorkspaceContextDeliveryRecord = z.infer<typeof workspaceContextDeliveryRecordSchema>

export const hostResponseKindSchema = z.enum([
  'host_fact',
  'tool_result',
  'delegation_acknowledgement',
])

export const hostResponseIntentSchema = z.object({
  kind: hostResponseKindSchema,
  item: hostContextItemSchema,
  task_summary: boundedText().nullable().default(null),
  origin_spoken: z.boolean().default(false),
}).strict().superRefine((intent, context) => {
  if (intent.item.kind === 'workspace_context') {
    context.addIssue({
      code: 'custom',
      path: ['item'],
      message: 'workspace context cannot create a response',
    })
  }
  if (intent.kind === 'host_fact' && intent.item.kind === 'tool_output') {
    context.addIssue({code: 'custom', path: ['item'], message: 'host fact cannot be tool output'})
  }
  if (intent.kind === 'host_fact' && intent.item.kind === 'dialogue_context') {
    context.addIssue({
      code: 'custom',
      path: ['item'],
      message: 'dialogue context cannot create a response',
    })
  }
  if (intent.kind === 'tool_result' && intent.item.kind !== 'tool_output') {
    context.addIssue({code: 'custom', path: ['item'], message: 'tool result requires tool output'})
  }
  if (intent.kind === 'delegation_acknowledgement') {
    if (intent.item.kind !== 'tool_output') {
      context.addIssue({
        code: 'custom',
        path: ['item'],
        message: 'delegation acknowledgement requires tool output',
      })
    }
    if (intent.task_summary === null) {
      context.addIssue({code: 'custom', path: ['task_summary'], message: 'task_summary is required'})
    }
  } else {
    if (intent.task_summary !== null) {
      context.addIssue({
        code: 'custom',
        path: ['task_summary'],
        message: 'task_summary is only valid for delegation acknowledgement',
      })
    }
    if (intent.origin_spoken) {
      context.addIssue({
        code: 'custom',
        path: ['origin_spoken'],
        message: 'origin_spoken is only valid for delegation acknowledgement',
      })
    }
  }
})

export type HostResponseIntent = z.infer<typeof hostResponseIntentSchema>

export const sessionIdentitySchema = z.object({
  epoch: epochSchema,
  provider_session_id: realtimeIdentifierSchema,
}).strict()

export const itemIdentitySchema = z.object({
  session_epoch: epochSchema,
  host_item_id: realtimeIdentifierSchema,
  provider_item_id: realtimeIdentifierSchema,
}).strict()

export type SessionIdentity = z.infer<typeof sessionIdentitySchema>
export type ItemIdentity = z.infer<typeof itemIdentitySchema>

const sessionEvent = <Kind extends z.ZodLiteral<string>, Shape extends z.ZodRawShape>(
  kind: Kind,
  shape: Shape,
) => z.object({kind, session_epoch: epochSchema, ...shape}).strict()

const itemTextShape = {
  item_id: realtimeIdentifierSchema,
  text: boundedText(true),
}

const responseTextShape = {
  response_id: realtimeIdentifierSchema,
  text: boundedText(true),
}

export const userSpeechStartedSchema = sessionEvent(z.literal('user_speech_started'), {
  speech_id: realtimeIdentifierSchema,
  provider_item_id: realtimeIdentifierSchema.nullable().default(null),
})
export const userSpeechEndedSchema = sessionEvent(z.literal('user_speech_ended'), {
  speech_id: realtimeIdentifierSchema,
  provider_item_id: realtimeIdentifierSchema.nullable().default(null),
})
export const userTranscriptDeltaSchema = sessionEvent(
  z.literal('user_transcript_delta'),
  itemTextShape,
)
export const userTranscriptFailedSchema = sessionEvent(z.literal('user_transcript_failed'), {
  item_id: realtimeIdentifierSchema,
})
export const userTranscriptFinalSchema = sessionEvent(
  z.literal('user_transcript_final'),
  itemTextShape,
)
export const responseStartedSchema = sessionEvent(z.literal('response_started'), {
  response_id: realtimeIdentifierSchema,
})
/**
 * Inbound provider audio is bounded by alignment, not by size.
 *
 * `ResponseAudioDelta.__post_init__` requires only non-empty PCM16 alignment, and the playback
 * registry is built for larger deltas: it splits one into `MAX_PLAYBACK_FRAME_BYTES` frames. A
 * size bound here would make that split unreachable from the provider path and would refuse audio
 * the oracle accepts -- narrowing the accepted domain on one leg only. `MAX_REALTIME_PCM_BYTES`
 * still bounds what we *send*, which is a different direction with its own reason.
 */
export const responseAudioDeltaSchema = sessionEvent(z.literal('response_audio_delta'), {
  response_id: realtimeIdentifierSchema,
  pcm: z.instanceof(Uint8Array)
    .refine(value => value.byteLength > 0 && value.byteLength % 2 === 0,
      'pcm must be non-empty aligned PCM16 bytes'),
})
export const responseTranscriptDeltaSchema = sessionEvent(
  z.literal('response_transcript_delta'),
  responseTextShape,
)
export const responseTranscriptFinalSchema = sessionEvent(
  z.literal('response_transcript_final'),
  responseTextShape,
)
export const toolCallReadySchema = sessionEvent(z.literal('tool_call_ready'), {
  call_id: realtimeIdentifierSchema,
  item_id: realtimeIdentifierSchema,
  name: realtimeIdentifierSchema,
  arguments: jsonObjectSchema,
  response_id: realtimeIdentifierSchema.nullable().default(null),
})
export const itemConfirmedSchema = sessionEvent(z.literal('item_confirmed'), {
  host_item_id: realtimeIdentifierSchema,
  provider_item_id: realtimeIdentifierSchema,
})
export const responseTerminalSchema = sessionEvent(z.literal('response_terminal'), {
  response_id: realtimeIdentifierSchema,
  status: z.enum(['completed', 'cancelled', 'failed']),
  reason: boundedText(),
})
export const responseCancelRejectedSchema = sessionEvent(z.literal('response_cancel_rejected'), {
  response_id: realtimeIdentifierSchema,
  cancel_request_id: realtimeIdentifierSchema,
  reason: z.literal('no_active_response'),
})
export const providerErrorEventSchema = sessionEvent(z.literal('provider_error'), {
  code: realtimeIdentifierSchema,
  recoverable: z.boolean().default(false),
})

export const realtimeProviderEventSchema = z.discriminatedUnion('kind', [
  userSpeechStartedSchema,
  userSpeechEndedSchema,
  userTranscriptDeltaSchema,
  userTranscriptFailedSchema,
  userTranscriptFinalSchema,
  responseStartedSchema,
  responseAudioDeltaSchema,
  responseTranscriptDeltaSchema,
  responseTranscriptFinalSchema,
  toolCallReadySchema,
  itemConfirmedSchema,
  responseTerminalSchema,
  responseCancelRejectedSchema,
  providerErrorEventSchema,
])

export type RealtimeProviderEvent = z.infer<typeof realtimeProviderEventSchema>
export type JsonObject = Readonly<Record<string, JsonValue>>

export interface RealtimeProvider {
  connect(options: {
    readonly tools: readonly JsonObject[]
    readonly signal: AbortSignal
  }): Promise<unknown>
  sendAudio(pcm: Uint8Array, signal: AbortSignal): Promise<void>
  injectHostItem(
    item: HostContextItem,
    options: {
      readonly confirmationTimeout: number | null
      readonly asUserActivation: boolean
      readonly signal: AbortSignal
    },
  ): Promise<unknown>
  injectWorkspaceContext?(
    item: HostContextItem,
    options: {
      readonly confirmationTimeout: number | null
      readonly signal: AbortSignal
    },
  ): Promise<unknown>
  createResponse(intent: HostResponseIntent, signal: AbortSignal): Promise<void>
  cancelResponse(responseId: string, signal: AbortSignal): Promise<void>
  events(signal: AbortSignal): AsyncIterable<unknown>
  close(): Promise<void>
}

export class RealtimeProtocolError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'RealtimeProtocolError'
  }
}

export class ItemDeliveryUncertainError extends Error {
  readonly session_epoch: number
  readonly host_item_id: string
  readonly provider_item_id: string
  readonly item_kind: z.infer<typeof hostItemKindSchema>

  constructor(input: {
    readonly session_epoch: number
    readonly host_item_id: string
    readonly provider_item_id: string
    readonly item_kind: z.infer<typeof hostItemKindSchema>
  }) {
    super('host item confirmation timed out; delivery is uncertain')
    const identity = itemIdentitySchema.parse({
      session_epoch: input.session_epoch,
      host_item_id: input.host_item_id,
      provider_item_id: input.provider_item_id,
    })
    this.name = 'ItemDeliveryUncertainError'
    this.session_epoch = identity.session_epoch
    this.host_item_id = identity.host_item_id
    this.provider_item_id = identity.provider_item_id
    this.item_kind = hostItemKindSchema.parse(input.item_kind)
  }
}

export function hostFact(item: HostContextItem): HostResponseIntent {
  return hostResponseIntentSchema.parse({kind: 'host_fact', item})
}

export function toolResult(item: HostContextItem): HostResponseIntent {
  return hostResponseIntentSchema.parse({kind: 'tool_result', item})
}

export function delegationAcknowledgement(
  item: HostContextItem,
  taskSummary: string,
  originSpoken = false,
): HostResponseIntent {
  return hostResponseIntentSchema.parse({
    kind: 'delegation_acknowledgement',
    item,
    task_summary: taskSummary,
    origin_spoken: originSpoken,
  })
}
